BaristaSwift/Common/Common/Base/Store.swift

```diff
@@ -89,95 +88,87 @@ public class Store<objectType: StoreProtocol> {
-        var record: T = entry
-        record.local = true
-        record.locallyUpdated = true
-        syncEntry(entry: record, completion: completion)
+    public func deleteEntry(entry: objectType) -> Promise<Void> {
+        locallyDeleteEntry(entry: entry)
+        return syncUpDown()
     }
 
-    public func syncEntry<T:StoreProtocol>(entry: T, completion: SyncCompletion = nil) {
-        var record: T = entry
-        record.objectType = T.objectName
-        store.upsertEntries([record.data], toSoup: T.objectName)
-        syncUp() { syncState in
-            if let _ = syncState?.isDone() {
-                self.syncDown(completion: completion)
-            }
-        }
+    public func syncEntry(entry: objectType) -> Promise<Void> {
+        var record: objectType = entry
+        record.objectType = objectType.objectName
+        store.upsertEntries([record.data], toSoup: objectType.objectName)
+        return syncUpDown()
     }
 
-    public func syncDown(completion: SyncCompletion = nil) {
-        smartSync.reSync(byName: syncDownName, update: completion ?? { _ in return })
-    }
-    
-    public func syncUp(completion: SyncCompletion = nil) {
-        let updateBlock: SFSyncSyncManagerUpdateBlock = { [unowned self] (syncState: SFSyncState?) in
-            if let syncState = syncState {
-                if syncState.isDone() || syncState.hasFailed() {
-                    DispatchQueue.main.async {
-                        if syncState.hasFailed() {
-                            SalesforceSwiftLogger.log(type(of:self), level:.error, message:"syncUp \(objectType.objectName) failed")
-                        }
-                    }
-                    completion?(syncState)
+    private func reSync(syncName: String) -> Promise<Void> {
+        return smartSync.Promises.reSync(syncName: syncName)
+            .then { syncState -> Promise<Void> in
+                if syncState.hasFailed() {
+                    SalesforceSwiftLogger.log(type(of:self), level:.error, message:"sync \(syncName) failed")
+                    return Promise(error:StoreErrors.syncFailed)
                 }
-            }
+                SalesforceSwiftLogger.log(type(of:self), level:.error, message:"sync \(syncName) completed")
+                return Promise.value(())
         }
-        
-        DispatchQueue.main.async(execute: {
-            self.smartSync.reSync(byName: self.syncUpName, update: updateBlock)
-        })
     }
     
-    public func syncUpDown(completion: SyncCompletion) {
-        self.syncUp { (upState) in
-            if let upComplete = upState?.isDone(), upComplete == true {
-                self.syncDown(completion: completion)
-            }
-        }
+    public func syncDown() -> Promise<Void> {
+        return reSync(syncName: syncDownName)
+    }
+    
+    public func syncUp() -> Promise<Void> {
+        return reSync(syncName: syncUpName)
```

Comment:
Shouldn't this be `syncUp(syncName: syncUpName)`?

Reply:
reSync(syncName:x) will do a full sync the first time it's run and a delta sync subsequent times and it can be used for sync down and sync up.